docs(tutorial): serverless triggers&functions setup & deploy MTA-3308 (#1835)
scaleway/docs-content: tutorials/transform-bucket-images-triggers-functions-set-up/index.mdx

---
meta:
  title: Transforming images in an S3 bucket using Serverless Functions and Triggers - Set up
  description: This page shows you how to set up your environment to transform images in an S3 bucket using Serverless Functions and Triggers
content:
  h1: Transforming images in an S3 bucket using Serverless Functions and Triggers - Set up
  paragraph: This page shows you how to set up your environment to transform images in an S3 bucket using Serverless Functions and Triggers
categories:
  - serverless
  - object storage
tags: serverless serverless-functions functions serverless-triggers triggers sqs-queue bucket
dates:
  validation: 2023-07-07
  posted: 2023-07-07
---

## Introduction

This tutorial is the first of a series of two. You will learn how to set up your environment, [then how to create and deploy functions](/tutorials/create-deploy-serverless-functions/) to resize images stored in a bucket and push them to another bucket.

<Lightbox src="scaleway-faas-image-transform.webp" alt="" />

Serverless Functions are an asynchronous microservices architecture where event sources are decorrelated from event consumers.

> They work best when they are triggered by events stored in queues, which means that they can be edited without having to modify every microservices in the process.

Reviewer: ```suggestion
They work best when they are triggered by events stored in queues, which means that they can be edited without having to modify every microservice in the process.
```

In this tutorial, you will create the following resources to use your functions:

- A project dedicated to this tutorial
- Two buckets to store media
- A queue to manage events
- IAM credentials to secure your environment
- Cockpit to monitor your functions

<Macro id="iam-requirements" />
 
<Message type="requirement">
  - You have an account and are logged into the [Scaleway console](https://console.scaleway.com).
  - You have around 10 JPG or PNG images.
  - You have [Vim](https://www.vim.org/download.php) installed.
</Message>

## Creating a New Project

1. Click the **Projects** tab from the [Organization Dashboard](https://console.scaleway.com/organization). A list of your Projects displays.

> 2. Click **+ Create Project** on the righthand side. A pop-up appears. 

Reviewer: ```suggestion
2. Click **+ Create Project** on the right-hand side. A pop-up appears. 
```

> 2. Click **+ Create Project** on the righthand side. A pop-up appears. 

Reviewer 2: ```suggestion
2. Click **+ Create Project** on the righthand side. A pop-up displays. 
```

3. Name the project "Functions Tutorial".
4. Click **Create Project**. The new Project is added to the list. 

## Creating the source bucket

1. Click **Object Storage** under **Storage** on the left side menu of the console. The Object Storage dashboard displays.
3. Click **+ Create a bucket**.
4. Name it "source-images-<YOUR_NAME>".

> 5. Chose the **AMSTERDAM** Region.

Reviewer: ```suggestion
5. Choose the **AMSTERDAM** Region.
```

6. Set the bucket visibility to **Private**.
7. Click **Create Bucket**. The bucket **Files** tab displays.
8. Upload your images to the bucket and select the **Standard** storage class.

## Creating the destination bucket

1. Click **Storage**, then **Object Storage** on the left side menu of the console. The Object Storage dashboard displays.
2. Click **+ Create a bucket**.

> 3. Name it "dest-images-<YOUR_NAME>".

Reviewer 3: ```suggestion
3. Name it `dest-images-<YOUR_NAME>`.
```
in the future, you can also just add a link to the create a bucket procedure to avoid duplicates :D

> 4. Chose the **AMSTERDAM** Region.

Reviewer: ```suggestion
4. Choose the **AMSTERDAM** Region.
```

5. Set the bucket visibility to **Private**.
6. Click **Create Bucket**.

> ## Creating a messaging and queuing namespace

Reviewer 2: ```suggestion
## Creating a Messaging and Queuing namespace
```

not sure about this one, but as M&Q is a product name

Message queues allow functions to communicate with each other.

1. Click **Messaging** under **Serverless**.
2. Click **Create namespace**.
3. Choose **SQS/SNS** as type.
4. Choose a name or generate one.
5. Click **Create namespace**.

## Generating credentials for the M&Q namespace

1. Click on the namespace you just created.
2. Click **Generate credentials**.

> 3. Name your credentials "Function Push".

Reviewer 3: ```suggestion
3. Name your credentials `Function Push`.
```

4. Enable the **Write**, **Read** and **Manage** permissions.
5. Click **Generate credentials**.

> 6. Store the Access key ID and the Secret key securely, as you will not be able to access them later.

Reviewer 3: ```suggestion
6. Store the access key ID and the secret key securely, as you will not be able to access them later.
```

## Creating a Queue

1. Click the **Queues** tab of your namespace.
2. Click **Create queue**.

> 3. Select *Standard** as a queue type.

Reviewer 3: ```suggestion
3. Select **Standard** as a queue type.
```

> 4. Name it "image-queue". 

Reviewer 3: ```suggestion
4. Name it `image-queue`. 
```

5. Keep the default parameters.
6. Click **Create queue**.

## Creating IAM credentials

> 

Reviewer 3: same here, you can maybe just add a link to the procedure to make tuto a bit shorter, but it's ultimately up to you!

Author: We need to force some options (Set up preferred project for object storage, User as a bearer and 1 week expiration, name) for this tutorial.  IAM How-to leaves the choice open, which can lead to tutorial not working. Also, numbers of steps performed remains the same for the user, so might as well keep it now that it is written!

Identity and Access Management (IAM) credentials allow you to secure access to your buckets, functions, etc. 

<Message type="note">

>     For this tutorial we will use *User* credentials which are simpler to use. However for production-grade Serverless applications, we recommend to use *Application* credentials with the minimum possible rights.

Reviewer: ```suggestion
    For this tutorial we will use *User* credentials which are simpler to use. However, for production-grade Serverless applications, we recommend to use *Application* credentials with the minimum possible rights.
```

</Message>

> 1. Click the Organization on the top right corner and Select **API Keys**. The IAM **API keys** tab displays.

Reviewer: ```suggestion
1. Click the Organization in the top right corner and Select **API Keys**. The IAM **API keys** tab displays.
```

2. Click **+ Generate an API key** or the <Icon name="plus" /> button.
3. Select **Myself (IAM)** as the API Key Bearer.

> 4. Enter “API for the Function Tutorial” as a description.

Reviewer 3: ```suggestion
4. Enter `API for the Function Tutorial` as a description.
```

5. Set **Expiration** to **One Week**.
6. Select **Yes, set up preferred Project**
7. Choose the **Functions Tutorial** project (or the name you gave to this project).
8. Click **Generate API Key**.
9. Store the Access key ID and the Secret key securely, as you will not be able to access them later.

## Enabling monitoring with Cockpit

1. Click **Cockpit** under **Observability**.
2. Click **Activate Cockpit**.
3. Click the **Grafana users** tab. 
4. Click **+ Add user**.
5. Enter a username.
6. Select the **Editor** role.
7. Click **Add user**.
8. Store the password securely, as you will not be able to access it later.

## Conclusion

You have now created all the resources to use Serverless Functions.

> Setting up an environment with all the necessary resources can be time consuming. You can use IaC (Infrastructure as Code) tools such as [Terraform](/tutorials/manage-instances-with-terraform-and-functions/) to automate a project's configuration.

Reviewer: ```suggestion
Setting up an environment with all the necessary resources can be time-consuming. You can use IaC (Infrastructure as Code) tools such as [Terraform](/tutorials/manage-instances-with-terraform-and-functions/) to automate a project's configuration.
```
